scripts/fetch_data.py

```diff
@@ -0,0 +1,150 @@
+import logging
+import sqlite3
+import time
+
+import numpy as np
+import pandas as pd
+from sqlalchemy import create_engine
+from sqlalchemy.exc import OperationalError, IntegrityError
+
+from lib.constants import SAVE_DIR, DATA_DIR
+from lib.data import (
+    fetch_physical_data,
+    add_bm_unit_type,
+    parse_boal_from_physical_data,
+    parse_fpn_from_physical_data,
+)
+from lib.db_utils import drop_and_initialize_tables
+
+engine = create_engine("sqlite:///phys_data.db", echo=False)
+df_bm_units = pd.read_excel(DATA_DIR / "BMUFuelType.xls", header=0)
+
+logging.basicConfig(level=logging.INFO)
+
+MAX_RETRIES = 1
+
+
+def run(start_date, end_date, units, chunk_size_in_days=7):
+    """
+    Collects data from the ElexonAPI, saved as a local feather file, does some preprocessing and then places in
+    an SQLite DB.
+
+    Only collects data for specified units, to keep things fast. Uses multiprocessing to grab all units in parallel.
+    """
+
+    interval = pd.Timedelta(days=chunk_size_in_days)
+
+    chunk_start = start_date
+    chunk_end = start_date + interval
+
+    while chunk_end < end_date:
+        logger.info(f"{chunk_start} to {chunk_end}")
+        t1 = time.time()
+        fetch_and_load_one_chunk(start_date=str(chunk_start), end_date=str(chunk_end), unit_ids=units)
+        t2 = time.time()
+        logger.info(f"{(t2 - t1) / 60} minutes for {interval}")
+        chunk_start = chunk_end
+        chunk_end += interval
+
+
+def write_fpn_to_db(df_fpn) -> bool:
+    """Write the FPN df to DB"""
+    try:
+        with engine.connect() as connection:
+            df_fpn.to_sql("fpn", connection, if_exists="append", index_label="unit")
+        return True
+    except OperationalError:
+        return False
+
+
+def write_boal_to_db(df_boal) -> bool:
+    """Write the BOAL df to DB, falling back to a row-by-row load if the load of the whole df fails.
+
+    This can happen because at boundaries between SPs, the same BOAL can be reported in multiple SP's. For instance,
+    if the BOAL is 00:40 -> 01.05, it will be reported in two SPs, and so we can end up trying to load the same
+    BOAL twice.
+    """
+    try:
+        with engine.connect() as connection:
+            # Potential issue here with duplicate BOALs nuking the whole write. This can happen because
```

Comment:
What do you mean by 'nuking' the whole thing?

Reply:
Because the `df` is written in one line, if any of the rows fail the whole write fails.